### PR TITLE
Explorer txn client

### PR DESCRIPTION
**Pull request by GusevTimofey**

### CODE REVIEW

File: cardano-tx/src/CardanoTx/Models.hs [hunk 1]
```diff
@@ -142,3 +143,36 @@ data TxCandidate = TxCandidate
...
+
+data CompletedRedeemer = CompletedRedeemer
+  { redeemer   :: Maybe Redeemer
+  , scriptHash :: Hash28
+  } deriving (Show, Eq, Generic, FromJSON)
+
+data CompletedTxIn = CompletedTxIn
+  { out      :: FullTxOut
+  , redeemer :: Maybe CompletedRedeemer
+  } deriving (Show, Eq, Generic, FromJSON)
```
Comment (oskin1) [inline cardano-tx/src/CardanoTx/Models.hs:170]: Ledger.TxIn contains this data already

File: cardano-tx/src/CardanoTx/Models.hs [hunk 2]
```diff
@@ -142,3 +143,36 @@ data TxCandidate = TxCandidate
...
+
+data CompletedTx = CompletedTx
+  { blockHash  :: BlockHash
+  , blockIndex :: Int
```
Comment (oskin1) [inline cardano-tx/src/CardanoTx/Models.hs:174]: What is block index on protocol level? please make the naming more clear

File: cardano-tx/src/CardanoTx/Models.hs [hunk 3]
```diff
@@ -142,3 +143,36 @@ data TxCandidate = TxCandidate
...
+
+-- TX hash32 represented as a hex string
+newtype TxHash = TxHash { unTxHash :: T.Text }
+  deriving (Eq, Generic)
+  deriving newtype (Show, FromJSON)
+
+-- Block hash32 represented as a hex string
+newtype BlockHash = BlockHash { unBlockHash :: T.Text }
+  deriving (Eq, Generic)
+  deriving newtype (Show, FromJSON)
```
Comment (oskin1) [inline cardano-tx/src/CardanoTx/Models.hs:160]: Please move types to Types.hs

**Review by oskin1 — CHANGES_REQUESTED** [review]

File: cardano-tx/src/CardanoTx/Models.hs [hunk 4]
```diff
@@ -142,3 +144,11 @@ data TxCandidate = TxCandidate
   , txCandidateValidRange   :: SlotRange
   , txCandidateSigners      :: [PaymentPubKeyHash]
   } deriving (Show, Eq, Generic, FromJSON, ToJSON)
+
+data CompletedTx = CompletedTx
```
Comment (oskin1) [inline cardano-tx/src/CardanoTx/Models.hs:148]:
```suggestion
data SettledTx = SettledTx
```

File: dex-core/src/ErgoDex/Class.hs [hunk 5]
```diff
@@ -8,3 +8,6 @@ class FromLedger a where
 
 class ToLedger a where
   toLedger :: a -> TxOutCandidate
+
+class FromExplorer a b where
+  parseFromExplorer :: a -> Maybe b
```
Comment (oskin1) [inline dex-core/src/ErgoDex/Class.hs:13]: Core module should not have any dependencies on explorer. Makes sense to put a TC like this into explorer module.

File: dex-core/src/ErgoDex/Index/ExecutedOrder.hs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+import           GHC.Generics                (Generic)
+
+data ExecutedSwap = ExecutedSwap
+  { config       :: Swap
+  , actualQuote  :: Amount Quote
+  , orderInputId :: String
+  , userOutputId :: String
+  , poolOutputId :: String
+  , poolInputId  :: String
+  } deriving (Show, Eq, Generic, FromJSON, ToJSON)
```
Comment (oskin1) [inline dex-core/src/ErgoDex/Index/ExecutedOrder.hs:24]: Why putting models like this to core lib?

**Review by oskin1 — COMMENTED** [review]
I think this code should be placed in the specific service, but not in dex-core.

File: cardano-tx/src/CardanoTx/Models.hs (hunk 4 above)
Comment (GusevTimofey) [reply cardano-tx/src/CardanoTx/Models.hs]: renamed

File: dex-core/src/ErgoDex/Class.hs (hunk 5 above)
Comment (GusevTimofey) [reply dex-core/src/ErgoDex/Class.hs]: moved into quickblue module

File: dex-core/src/ErgoDex/Index/ExecutedOrder.hs (hunk 6 above)
Comment (GusevTimofey) [reply dex-core/src/ErgoDex/Index/ExecutedOrder.hs]: moved into specific service

File: quickblue/src/Explorer/Models.hs [hunk 7]
```diff
@@ -74,6 +76,36 @@ data Items a = Items
...
+data FullTxIn = FullTxIn
+  { out :: FullTxOut
+  } deriving (Show, Generic, FromJSON)
+
+instance ToCardanoTx FullTxIn Tx.FullTxIn where
+  toCardanoTx FullTxIn{..} =
+    Tx.FullTxIn
+      { fullTxInTxOut = toCardanoTx out
+      -- actually, we don't need this field att all, so we keep it default
+      , fullTxInType  = Tx.ConsumeSimpleScriptAddress
```
Comment (Bromel777) [inline quickblue/src/Explorer/Models.hs:106]: why only ConsumeSimpleScriptAddress?

Comment (GusevTimofey) [reply quickblue/src/Explorer/Models.hs:106]: We do not need actual redeemer in our models (cause it contains just indexes of outs in tx)

Comment (GusevTimofey) [reply quickblue/src/Explorer/Models.hs:106]: We do not need this field at all, but model that already exists  contains it, so i put the simplest value by default

File: dex-core/src/ErgoDex/Index/ExecutedOrder.hs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        { redeemCfg          = cfg
+        , xReward            = assetAmountX
+        , yReward            = assetAmountY
+        , redeemOrderInputId = show P.$ fullTxOutRef redeemOut
+        , redeemUserOutputId = show P.$ fullTxOutRef userOutput
+        , currPool           = show P.$ fullTxOutRef currPool
+        , prevPoolId         = show P.$ fullTxOutRef prevPool
+        }
+
+findInput :: forall a . (FromLedger a) => [FullTxIn] -> Maybe (OnChain a)
```
Comment (Bromel777) [inline dex-core/src/ErgoDex/Index/ExecutedOrder.hs]: findInput -> findFirstInput?

File: dex-core/src/ErgoDex/Index/ExecutedOrder.hs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+findInput (FullTxIn{..}:xs) =
+  case parseFromLedger fullTxInTxOut of
+      Just r -> Just r
+      _      -> findInput xs
+findInput [] = Nothing
+
+findOutput :: forall a . (FromLedger a) => [FullTxOut] -> Maybe (OnChain a)
```
Comment (Bromel777) [inline dex-core/src/ErgoDex/Index/ExecutedOrder.hs]: findOutput -> findFirstOutput?

**Review by oskin1 — APPROVED** [review]

**Pull request merged**